src/game_set.ts

```diff
@@ -32,34 +36,244 @@ export default class GameSet extends Entity {
+      public angle: number;
+      public game_set: GameSet;
+      constructor(position: Vector2D, angle: number, game_set: GameSet) {
+        this.position = position;
+        this.angle = angle;
+        this.game_set = game_set;
+      }
+    };
+
+    const all_objects_except_me = this.filter_contents(o => o.id != object.id).map(e => e.id);
+
+    const final_delta = all_objects_except_me.reduce(
+      (delta, to_collide) => {
+        const updated_self = delta.game_set.contents.get(object.id) as PhysicalObject;
+        const updated_to_collide = delta.game_set.contents.get(to_collide) as PhysicalObject;
+        
+        const collision_result = this._collide(
+          updated_self,
+          updated_to_collide,
+          delta.position,
+          delta.angle,
+          delta.game_set);
 
-      const new_game_set = first_object.update_game_set(time_unit, game_set) as GameSet;
-      return updatedRec(new_game_set, remaining.shift());
+        const new_game_set = collision_result.game_set as GameSet;
+        const new_delta_position = collision_result.delta_position;
+        const new_delta_angle = collision_result.delta_angle;
+
+        return new DeltaState(new_delta_position, new_delta_angle, new_game_set);
+      },
+      new DeltaState(delta_position, delta_angle, this.replace_element(next_state_resetted_translation)));
+    
+      const updated_self = final_delta.game_set.contents.get(object.id);
+      const updated_self_with_delta = updated_self.move(final_delta.position).rotate(final_delta.angle);
+      const must_reset = final_delta.game_set.filter_contents(o => o.id != object.id).some(other =>
+          updated_self_with_delta.calculate_collision(other).collided());
+      
+      return must_reset ? final_delta.game_set : final_delta.game_set.replace_element(updated_self_with_delta);
+  }
+
+  public _collide(o_a: PhysicalObject, o_b: PhysicalObject, delta_position: Vector2D, delta_angle: number, game_set: GameSet): CollisionResult {
+    const advanced_self = o_a.move(delta_position).rotate(delta_angle);
+
+    const collision = advanced_self.calculate_collision(o_b);
+    if (!collision.collided()) {
+      return {
+        game_set: game_set,
+        delta_position: delta_position,
+        delta_angle: delta_angle
+      };
     }
 
-    const updated_game_set = updatedRec(this,
-        this.contents.entrySeq().filter(([k, v]) => !v.is_ground).map(([k, v]) => k).toList());
+    // Impulse-based collision handling.
+    // Reference: https://www.myphysicslab.com/engine2D/collision-en.html#collision_physics
+    const impulse_weight = 1.0 / collision.intersections.size;
+
+    if (o_b.is_ground) {
+      const collide_rec = (delta_v_a: Vector2D, delta_w_a: number, delta_p: Vector2D,
+                           remaining_intersections: Immutable.List<Intersection>): any => {
+        if (remaining_intersections.size == 0) {
+          return {
+            d_v_a: delta_v_a,
```

Comment:
So cryptic it's almost like minified code.
And given this function has `any` for return type, I literally have no idea what to make of this code. :)

Reply:
lol, can't disagree. I'm reworking + renaming variables in this function in a later PR though, tbh, Physics is cryptic!